File: tools/oracle/primitives.c
#include <stdio.h>

#include "../../src/microui.h"

int main(void) {
  mu_Vec2 v = mu_vec2(1, 2);
  mu_Rect r = mu_rect(3, 4, 5, 6);
  mu_Color c = mu_color(7, 8, 9, 10);

  printf("version=%s\n", MU_VERSION);
  printf("vec2=Vec2 { x: %d, y: %d }\n", v.x, v.y);
  printf("rect=Rect { x: %d, y: %d, w: %d, h: %d }\n", r.x, r.y, r.w, r.h);
  printf("color=Color { r: %u, g: %u, b: %u, a: %u }\n", c.r, c.g, c.b, c.a);
  return 0;
}

File: tests/primitives.rs
mod common;

use common::fixture;
use microui::{color, rect, vec2, Color, Rect, Vec2, MU_VERSION};

#[test]
fn constructors_and_version_exist() {
    assert_eq!(MU_VERSION, "2.02");
    assert_eq!(vec2(1, 2), Vec2 { x: 1, y: 2 });
    assert_eq!(rect(3, 4, 5, 6), Rect { x: 3, y: 4, w: 5, h: 6 });
    assert_eq!(color(7, 8, 9, 10), Color { r: 7, g: 8, b: 9, a: 10 });
}

#[test]
fn primitive_fixture_matches_c_oracle() {
    let fixture_text = fixture("tests/fixtures/primitives.txt");
    let actual = format!(
        "version={}\nvec2={:?}\nrect={:?}\ncolor={:?}\n",
        MU_VERSION,
        vec2(1, 2),
        rect(3, 4, 5, 6),
        color(7, 8, 9, 10),
    );
    assert_eq!(actual, fixture_text);
}
